ElrondNetwork/elrond-proxy-go
# EN-8539: observers live reloading and api endpoints configuration files

Implemented:

1. `api.toml` configuration file that contains individual settings for each endpoint (closed and secured)
2.  basic authentication for selected endpoints in `api.toml`
3.  two endpoints that enable 'live' reloading for observers and full history nodes  

Testing procedure:
1. `api.toml`
  1.1. edit `api.toml` and set any endpoint as closed or as secured. should act as expected

2. basic authentication
  2.1. define new credentials in `credentials.toml`. you should save pairs of `username - sha256(password)`
  2.2. make requests to secured endpoints. You should set `basic authentication` and add the `username - password` (in the request, the password is plain-text and in credentials.toml it is hashed). more details below

3. live reloading endpoints
  3.1. send a POST api call to `proxy/actions/reload-observers` with no body. Expected: fail with code 401 and error `this endpoint requires Basic Authentication`
  3.2. send a POST api to `proxy/actions/reload-observers` with no body and a fake username for `Basic authentication`. expected: fail with code 401 and error `username does not exist`
  3.3. send a POST api to `proxy/actions/reload-observers` with no body and a good username and a bad password for `Basic authentication`. expected: fail with code 401 and error `invalid password`
  3.4. update the observers part of `config.toml` file after the proxy started. send a POST api to `proxy/actions/reload-observers` with no body and a good username and a good password for `Basic authentication`. expected: codes 200, 400 or 500 depending on the configuration files. The number of shard should be the same as the original - you can only add/remove observers to the same shards as the original configuration (otherwise, an error in thrown). After the request is successful, make requests to proxy and check that the new observers are used.

Example cUrl for secured endpoints (for a `example2` - `example2` credentials pair). The Authorization field can be obtained by: `base64(username:password)` . `ZXhhbXBsZTI6ZXhhbXBsZTI=` == `example2:example2`
```
curl --request POST \
  --url http://127.0.0.1:8080/actions/reload-observers \
  --header 'Authorization: Basic ZXhhbXBsZTI6ZXhhbXBsZTI=' \
  --header 'Content-Type: application/json'
```
